kernel/comps/virtio/src/device/socket/manager.rs

```diff
@@ -0,0 +1,357 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{boxed::Box, sync::Arc, vec, vec::Vec};
+use core::{cmp::min, hint::spin_loop};
+
+use aster_frame::sync::SpinLock;
+use log::debug;
+
+use super::{
+    connect::{ConnectionInfo, DisconnectReason, VsockEvent, VsockEventType},
+    device::SocketDevice,
+    header::VsockAddr,
+};
+use crate::device::socket::error::SocketError;
+
+const PER_CONNECTION_BUFFER_CAPACITY: usize = 1024;
+
+pub struct VsockConnectionManager {
+    driver: Arc<SpinLock<SocketDevice>>,
+    connections: Vec<Connection>,
+    listening_ports: Vec<u32>,
+}
+
+impl VsockConnectionManager {
+    /// Construct a new connection manager wrapping the given low-level VirtIO socket driver.
+    pub fn new(driver: Arc<SpinLock<SocketDevice>>) -> Self {
+        Self {
+            driver,
+            connections: Vec::new(),
+            listening_ports: Vec::new(),
+        }
+    }
+
+    /// Returns the CID which has been assigned to this guest.
+    pub fn guest_cid(&self) -> u64 {
+        self.driver.lock().guest_cid()
+    }
+
+    /// Allows incoming connections on the given port number.
+    pub fn listen(&mut self, port: u32) {
+        if !self.listening_ports.contains(&port) {
+            self.listening_ports.push(port);
+        }
+    }
+
+    /// Stops allowing incoming connections on the given port number.
+    pub fn unlisten(&mut self, port: u32) {
+        self.listening_ports.retain(|p| *p != port)
+    }
+    /// Sends a request to connect to the given destination.
+    ///
+    /// This returns as soon as the request is sent; you should wait until `poll` returns a
+    /// `VsockEventType::Connected` event indicating that the peer has accepted the connection
+    /// before sending data.
+    pub fn connect(&mut self, destination: VsockAddr, src_port: u32) -> Result<(), SocketError> {
+        if self.connections.iter().any(|connection| {
+            connection.info.dst == destination && connection.info.src_port == src_port
+        }) {
+            return Err(SocketError::ConnectionExists);
+        }
+        let new_connection = Connection::new(destination, src_port);
+        self.driver.lock().request(&new_connection.info)?;
+        debug!("Connection requested: {:?}", new_connection.info);
+        self.connections.push(new_connection);
+        Ok(())
+    }
+
+    /// Sends the buffer to the destination.
+    pub fn send(
+        &mut self,
+        destination: VsockAddr,
+        src_port: u32,
+        buffer: &[u8],
+    ) -> Result<(), SocketError> {
+        let (_, connection) = get_connection(&mut self.connections, destination, src_port)?;
+
+        self.driver.lock().send(buffer, &mut connection.info)
+    }
+
+    /// Polls the vsock device to receive data or other updates.
+    pub fn poll(&mut self) -> Result<Option<VsockEvent>, SocketError> {
```

Comment:
Each poll can update multiple socket status, right? So it cannot return exact one event.